File: java/bottle-song/src/main/java/BottleSong.java
class BottleSong {

    String recite(int startBottles, int takeDown) {
        StringBuilder verses = new StringBuilder();

        for (int i = startBottles; i > startBottles - takeDown; i--) {

            verses.append("\n");
            verses.append(getVerse(i));
            verses.append("\n");
        }

        return verses.toString().stripLeading();
    }


    private String getVerse(int startBottles) {

        String startBottlesStr = getBottles(startBottles);
        String endBottlesStr = getBottles(startBottles - 1).toLowerCase();

        String sBottle = startBottles > 1 ? "bottles" : "bottle";
        String eBottle = startBottles - 1 == 1 ? "bottle" : "bottles";

        return String.format(
                "%s green %s hanging on the wall,\n" +
                "%s green %s hanging on the wall,\n" +
                "And if one green bottle should accidentally fall,\n" +
                "There'll be %s green %s hanging on the wall.",
                startBottlesStr, sBottle,
                startBottlesStr, sBottle,
                endBottlesStr, eBottle);
    }

    private String getBottles(int bottles) {

        return switch (bottles) {
            case 10 -> "Ten";
            case 9 -> "Nine";
            case 8 -> "Eight";
            case 7 -> "Seven";
            case 6 -> "Six";
            case 5 -> "Five";
            case 4 -> "Four";
            case 3 -> "Three";
            case 2 -> "Two";
            case 1 -> "One";
            case 0 -> "No";
            default -> throw new IllegalArgumentException();
        };
    }


}

File: elixir/bottle-song/lib/bottle_song.ex
defmodule BottleSong do
  @moduledoc """
  Handles lyrics of the popular children song: Ten Green Bottles
  """

  @bottle_mappings %{
    10 => {"Ten", "bottles", "nine", "bottles"},
    9 => {"Nine", "bottles", "eight", "bottles"},
    8 => {"Eight", "bottles", "seven", "bottles"},
    7 => {"Seven", "bottles", "six", "bottles"},
    6 => {"Six", "bottles", "five", "bottles"},
    5 => {"Five", "bottles", "four", "bottles"},
    4 => {"Four", "bottles", "three", "bottles"},
    3 => {"Three", "bottles", "two", "bottles"},
    2 => {"Two", "bottles", "one", "bottle"},
    1 => {"One", "bottle", "no", "bottles"},
  }

  @spec recite(pos_integer, pos_integer) :: String.t()
  def recite(start_bottle, take_down) do
    Enum.to_list(start_bottle..(start_bottle - take_down + 1))
    |> Enum.map(&verse/1)
    |> Enum.join("\n\n")
  end


  defp verse(verse_number) do
    {start_bottle, bottle, end_bottle, bootle_remaining} = @bottle_mappings[verse_number]

    """
    #{start_bottle} green #{bottle} hanging on the wall,
    #{start_bottle} green #{bottle} hanging on the wall,
    And if one green bottle should accidentally fall,
    There'll be #{end_bottle} green #{bootle_remaining} hanging on the wall.\
    """

  end

end
